### PR TITLE
[ONNX] Add support for onnx tensor index export

### PR DESCRIPTION
**Pull request by BowenBao**
Support exporting
* Standard tensor indexing like
```
x = torch.ones(4, 5)
ind = torch.tensor([0, 1])

return x[ind]
```
* [Advanced indexing](https://docs.scipy.org/doc/numpy/reference/arrays.indexing.html#advanced-indexing) like
```
x = torch.ones(4,5,6,7,8)
ind1 = torch.tensor([0, 1])
ind2 = torch.tensor([[3], [2]])
ind3 = torch.tensor([[2, 2], [4, 5]])

return x[2:4, ind1, None, ind2, ind3, :]
```
It would be ideal if ONNX can natively support indexing in future opsets, but for opset <= 10 it will always need this kind of workarounds. 

There are still various limitations, such as not supporting advanced indexing with negative indices, not supporting mask indices of rank > 1, etc. My feeling is that these are less common cases that requires great effort to support using current opset, and it's better to not make the index export more cumbersome than it already is.

### CODE REVIEW

**Review by houseroad — COMMENTED** [review]
First, fix the linter?

File: torch/onnx/symbolic_opset9.py [hunk 1]
```diff
@@ -1612,3 +1612,109 @@ def gather(g, self, dim, index, sparse_grad=False):
     index = g.op("Cast", g.op("OneHot", index, depth, values, axis_i=dim), to_i=sym_help.cast_pytorch_to_onnx[dtype])
     mul = g.op("Mul", g.op("Unsqueeze", self, axes_i=[dim + 1]), index)
     return g.op("ReduceSum", mul, axes_i=[dim], keepdims_i=0)
+
+
+def index(g, self, index):
+    if sym_help._is_packed_list(index):
```
Comment (houseroad) [inline torch/onnx/symbolic_opset9.py]: The symbolic of index op incurs some internal pipe breakage.
The breakage is caused by the Nonzero operator. We don't have it in Caffe2.

At least we have two options so far:
1) implement the Nonzero op in Caffe2.
2) or add check here, if the export mode is ONNX_ATEN_FALLBACK, we direct emit aten operator

File: torch/onnx/symbolic_opset9.py [hunk 2]
```diff
@@ -1612,3 +1612,109 @@ def gather(g, self, dim, index, sparse_grad=False):
...
+        indices = sym_help._unpack_list(index)
+    else:
+        indices = [index]
+
+    def try_mask_to_index(index):
+        if not index.node().mustBeNone() and index.type().scalarType() == "Byte":
+            warnings.warn("Exporting aten::index operator with indices of type Byte. "
+                          "Only 1-D indices are supported. In any other case, "
+                          "this will produce an incorrect ONNX graph.")
+            index = squeeze(g, nonzero(g, index), dim=1)
```
Comment (houseroad) [inline torch/onnx/symbolic_opset9.py]: also seems our unittest didn't trigger the Nonzero part. Otherwise, the ci shoudl fail.

Comment (BowenBao) [reply torch/onnx/symbolic_opset9.py]: Good catch, adds a test case that uses nonzero.

File: torch/onnx/symbolic_opset9.py (hunk 1 above)
Comment (BowenBao) [reply torch/onnx/symbolic_opset9.py]: Thanks for the suggestions! Updated the PR to extend Caffe2 backend NonZero with ATen::nonzero.

**Review by houseroad — APPROVED** [review]
Looks good. Could you rebase again?

Btw, we should definitely introduce new index operator in ONNX opset 11. This translation is very complicated...

File: caffe2/onnx/backend.cc [hunk 3]
```diff
@@ -598,6 +599,30 @@ Caffe2Ops Caffe2Backend::CreateWhereOp(
   return CommonOnnxNodeToCaffe2Ops(&new_node, ctx);
 }
 
+Caffe2Ops Caffe2Backend::CreateNonZeroOp(
+    OnnxNode* onnx_node,
+    const ConversionContext& ctx) {
+  // Native Caffe2 doesn't support NonZero, fallback to ATen.
+  // ATen nonzero is equivalent to Transpose(ONNX::NonZero).
+  const auto& node = onnx_node->node;
```
[the remark this answers is not in the document]
Comment (BowenBao) [reply caffe2/onnx/backend.cc]: thanks

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 4]
```diff
@@ -154,6 +154,30 @@ def forward(self, x):
...
+                return fn(input)
+
+        m1 = torch.randn(3, 4, 5, 6, 7)
+        self.run_test(MyModel(), m1)
+
+    def test_tensor_index_advanced_indexing(self):
+        self._test_index_generic(
+            lambda input: input[:, torch.tensor([[0, 2], [1, 1]]), :, torch.tensor([2, 1]), torch.tensor([0, 3])])
+
+    def test_tensor_index_advanced_indexing2(self):
```
[the remark this answers is not in the document]
Comment (BowenBao) [reply test/onnx/test_pytorch_onnx_onnxruntime.py]: thanks, updated

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 5]
```diff
@@ -154,6 +154,30 @@ def forward(self, x):
...
+        self._test_index_generic(lambda input: input[:, torch.tensor([0, 2]), None, 2:4, torch.tensor([[1, 3], [4, 0]])])
+
+    def test_tensor_index_advanced_indexing3(self):
```
[the remark this answers is not in the document]
Comment (BowenBao) [reply test/onnx/test_pytorch_onnx_onnxruntime.py]: thanks, updated

File: torch/onnx/symbolic_helper.py [hunk 6]
```diff
@@ -122,6 +122,11 @@ def _unpack_list(list_value):
     return list(list_node.inputs())
 
 
+def _is_packed_list(list_value):
+    list_node = list_value.node()
```
[the remark this answers is not in the document]
Comment (BowenBao) [reply torch/onnx/symbolic_helper.py]: thanks, updated
